src/librustc_resolve/lib.rs

```diff
@@ -3461,6 +3524,103 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
         found_traits
     }
 
+    // Looks up a definitions (and paths) based on a user-supplied predicate
+    // This method should be used to generate suggestions when name resolution
+    // fails.
+    fn lookup_suggestions<FilterFn>(&mut self,
+                                    lookup_name: Name,
+                                    lookup_type: LookupTarget,
+                                    namespace: Namespace,
+                                    filter_fn: FilterFn) -> Vec<Path>
+        where FilterFn: Fn(Def) -> bool {
+
+        let mut lookup_results: Vec<Path> = Vec::new();
+        let mut lookup_modules: Vec<(Module, Vec<PathSegment>)> = Vec::new();
+        lookup_modules.push((self.graph_root, Vec::new()));
+
+        while !lookup_modules.is_empty() {
```

Comment:
`while let Some((in_module, path_segments)) = lookup_modules.pop()` saves you the `expect("???")` call